Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs

```diff
@@ -0,0 +1,455 @@
+
+        // Get the scroll position of the selected entity on the selected button the UI
+        _tryToScroll = true;
+
+        // Request an update from the power monitoring system
+        RequestPowerMonitoringUpdateAction?.Invoke(_entManager.GetNetEntity(_trackedEntity));
+        _updateTimer = 0f;
+    }
+
+    private bool TryGetNextScrollPosition([NotNullWhen(true)] out float? nextScrollPosition)
+    {
+        nextScrollPosition = null;
+
+        var scroll = MasterTabContainer.Children.ElementAt(MasterTabContainer.CurrentTab) as ScrollContainer;
+        if (scroll == null)
+            return false;
+
+        var container = scroll.Children.ElementAt(0) as BoxContainer;
+        if (container == null || !container.Children.Any())
+            return false;
+
+        // Exit if the heights of the children haven't been initialized yet
+        if (!container.Children.Any(x => x.Height > 0))
+            return false;
+
+        nextScrollPosition = 0;
+
+        foreach (var control in container.Children)
+        {
+            if (control == null || control is not PowerMonitoringWindowEntry)
+                continue;
+
+            if (((PowerMonitoringWindowEntry) control).EntityUid == _trackedEntity)
+                return true;
+
+            nextScrollPosition += control.Height;
+        }
+
+        // Failed to find control
+        nextScrollPosition = null;
+
+        return false;
+    }
+
+    private bool TryGetVerticalScrollbar(ScrollContainer scroll, [NotNullWhen(true)] out VScrollBar? vScrollBar)
+    {
+        vScrollBar = null;
+
+        foreach (var child in scroll.Children)
+        {
+            if (child is not VScrollBar)
+                continue;
+
+            var castChild = child as VScrollBar;
+
+            if (castChild != null)
+            {
+                vScrollBar = castChild;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void TryToScrollToFocus()
+    {
+        if (!_tryToScroll)
+            return;
+
+        var scroll = MasterTabContainer.Children.ElementAt(MasterTabContainer.CurrentTab) as ScrollContainer;
+        if (scroll == null)
+            return;
+
+        if (!TryGetVerticalScrollbar(scroll, out var vScrollbar))
+            return;
+
+        if (TryGetNextScrollPosition(out float? nextScrollPosition))
+        {
+            vScrollbar.ValueTarget = nextScrollPosition.Value;
+
+            if (MathHelper.CloseToPercent(vScrollbar.Value, vScrollbar.ValueTarget))
+            {
+                _tryToScroll = false;
+                return;
+            }
+        }
+    }
+
+    private void UpdateWarningLabel(PowerMonitoringFlags flags)
+    {
+        if (flags == PowerMonitoringFlags.None)
+        {
+            SystemWarningPanel.Visible = false;
+            return;
+        }
+
+        var msg = new FormattedMessage();
+
+        if (flags.HasFlag(PowerMonitoringFlags.RoguePowerConsumer))
+        {
```

Comment:
.HasFlag allocates in debug so we just normally manually bitmask it instead.

Reply:
Done